editor/plugins/skeleton_3d_editor_plugin.cpp

```diff
@@ -692,16 +1063,374 @@ void EditorInspectorPluginSkeleton::parse_begin(Object *p_object) {
+	}
+}
+
 Skeleton3DEditorPlugin::Skeleton3DEditorPlugin(EditorNode *p_node) {
 	editor = p_node;
 
-	Ref<EditorInspectorPluginSkeleton> skeleton_plugin;
-	skeleton_plugin.instantiate();
+	skeleton_plugin = memnew(EditorInspectorPluginSkeleton);
 	skeleton_plugin->editor = editor;
 
 	EditorInspector::add_inspector_plugin(skeleton_plugin);
+
+	Ref<Skeleton3DGizmoPlugin> gizmo_plugin = Ref<Skeleton3DGizmoPlugin>(memnew(Skeleton3DGizmoPlugin));
+	Node3DEditor::get_singleton()->add_gizmo_plugin(gizmo_plugin);
+}
+
+EditorPlugin::AfterGUIInput Skeleton3DEditorPlugin::forward_spatial_gui_input(Camera3D *p_camera, const Ref<InputEvent> &p_event) {
+	if (Skeleton3DEditor::get_singleton()->is_edit_mode()) {
+		const Ref<InputEventMouseButton> mb = p_event;
+		if (mb.is_valid() && mb->get_button_index() == MOUSE_BUTTON_LEFT) {
+			if (mb->is_pressed()) {
+				Skeleton3DEditor::get_singleton()->update_bone_original();
+			}
+		}
+		return EditorPlugin::UNSELECT;
+	}
+	return EditorPlugin::THROUGH;
+}
+
+bool Skeleton3DEditorPlugin::handles(Object *p_object) const {
+	return p_object->is_class("Skeleton3D");
+}
+
+void Skeleton3DEditor::_update_sub_gizmo() {
+	update_gizmo_visible();
+	Node3DEditor::get_singleton()->update_transform_gizmo();
+};
+
+void Skeleton3DEditor::update_gizmo_visible() {
+	if (edit_mode) {
+		if (skeleton->get_selected_bone() == -1) {
+			Node3DEditor::get_singleton()->set_gizmo_visible(false);
+		} else {
+			Node3DEditor::get_singleton()->set_gizmo_visible(true);
+		}
+	} else {
+		Node3DEditor::get_singleton()->set_gizmo_visible(true);
+	}
+}
+
+Skeleton3DGizmoPlugin::Skeleton3DGizmoPlugin() {
+	skeleton_color = EDITOR_DEF("editors/3d_gizmos/gizmo_colors/skeleton", Color(1, 0.8, 0.4));
+	selected_bone_color = EDITOR_DEF("editors/3d_gizmos/gizmo_colors/selected_bone", Color(0.8, 0.3, 0.0));
+	bone_axis_length = EDITOR_DEF("editors/3d_gizmos/gizmo_settings/bone_axis_length", (float)0.015);
+	create_material("skeleton_material", skeleton_color);
+	selected_mat = Ref<ShaderMaterial>(memnew(ShaderMaterial));
+	selected_sh = Ref<Shader>(memnew(Shader));
+	selected_sh->set_code(" \
+		shader_type spatial; \
+		render_mode unshaded, shadows_disabled; \
+		uniform vec4 albedo : hint_color = vec4(1,1,1,1); \
+		uniform sampler2D texture_albedo : hint_albedo; \
+		void vertex() { \
+			if (!OUTPUT_IS_SRGB) { \
+				COLOR.rgb = mix( pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb* (1.0 / 12.92), lessThan(COLOR.rgb,vec3(0.04045)) ); \
+			} \
+			VERTEX = VERTEX; \
+			POSITION=PROJECTION_MATRIX*INV_CAMERA_MATRIX*WORLD_MATRIX*vec4(VERTEX.xyz,1.0); \
+			POSITION.z = mix(POSITION.z, 0, 0.998); \
+		} \
+		void fragment() { \
+			vec2 base_uv = UV; \
+			vec4 albedo_tex = texture(texture_albedo,base_uv); \
+			albedo_tex *= COLOR; \
+			ALBEDO = albedo.rgb * albedo_tex.rgb; \
+			if (albedo.a * albedo_tex.a < 0.5) { discard; } \
+			ALPHA = albedo.a * albedo_tex.a; \
+		} \
+	");
```

Comment:
This whole shader should be simplified, the `vertex()` code is correct, but we don't need any of the uniforms and the fragment color just needs to set the ALBEDO. Something like this is probably enough: 

```
void fragment() {
    ALBEDO = COLOR;
}
```

Reply:
Related to the hack of the drawing order by Vertex, it is necessary to put it on Alpha's rendering pipeline to prevent Gizmo from casting shadows on the objects. So, finally
```
void fragment() { \
	ALBEDO = COLOR.rgb; \
	ALPHA = COLOR.a; \
} \
```